text/0101-revamping-the-sdk-performance-api.md

```diff
@@ -0,0 +1,273 @@
+transaction.finish();
+Sentry.getCurrentHub().getScope().setSpan(undefined);
+
+// second span info is not sent to Sentry because transaction is already finished.
+secondSpan.finish();
+```
+
+In our integrations that add automatic instrumentation, things look something like so:
+
+```jsx
+const parentSpan = Sentry.getCurrentHub().getSpan();
+
+// parentSpan can be undefined if no span is on the scope, this leads to
+// child span just being lost
+const span = parentSpan?.startChild({ description: "something" });
+
+Sentry.getCurrentHub().getScope().setSpan(span);
+
+work();
+
+span.finish();
+
+// span is finished, so parent is put back onto scope
+Sentry.getCurrentHub().getScope().setSpan(parentSpan);
+```
+
+Most users do the above also when nested within their application, as often you assume that a transaction is defined that you can attach (very common in a web server context).
+
+To add instrumentation to their applications, users have to know concepts about hubs/scopes/transactions/spans and understand all the different nuances and use cases. This can be difficult, and presents a big barrier to entry for new users.
+
+Also, since transactions/spans are different classes (span is a parent class of transaction), users have to understand the impacts that the same method will have on both the transaction/span. For example, currently calling `setTag` on a transaction will add a tag to the transaction event (which is searchable in Sentry), while `setTag` on a span just adds it to the span, and the field is not searchable. `setData` on a span adds it to `span.data`, while `setData` on a transaction is undefined behaviour (some SDKs throw away the data, others add it to `event.extras`).
+
+Summarizing, here are the core Issues in SDK Performance API:
+
+1. Users have to understand the difference between spans/transactions and their schema differences.
+2. Users have to set/get transactions/spans from a scope (meaning they also have to understand what a scope/hub means).
+3. Nesting transactions within each other is undefined behaviour - no obvious way to make a transaction a child of another.
+4. If a transaction finishes before it’s child span finishes, that child span gets orphaned and the data is never sent to Sentry. This is most apparent if you have transactions that automatically finish (like on browser/mobile).
+5. Transactions have a max child span count of 1000 which means that eventually data is lost if you keep adding child spans to a transaction.
+
+# Improvements
+
+## New SDK API
+
+The new SDK API has the following requirements:
+
+1. Newly created spans must have the correct trace and parent/child relationship
+2. Users shouldn’t be burdened with knowing if something is a span/transaction
+3. Spans only need a name to identify themselves, everything else is optional.
+4. The new top level APIs should be as similar to the OpenTelemetry SDK public API as possible.
```

Comment:
I'd be interested in that as well, what are the pro x cons of moving closer to opentelemetry rather than our own model?
I know a few of them but it'd be nice to be written down.